### ISSUE TITLE
How to test networks that takes 2D tensor as input

### ISSUE THREAD

**Issue by Rleahy** [SciSharp/TensorFlow.NET#243]
I have a network which takes a 2d tensor as an input I'm able to train the network by passing the input and output tensors in a loop as follows. 

```
                        for(int i=0;i<TrainingData.Count;i++)
                        {
                            sess.run( optimize, 
                                new FeedItem(ToInput, NDFromFloatArray(TrainingData[i].Inputs["ToFeatures"]) ),
                                new FeedItem(Output, TrainingData[i].Output) );
                            
                        }
```

This code executes correctly, However when it comes to testing the model on the validation set the examples show all values are passed in one array rather than one at a time in a loop.

The example I've found only take a single value so they are just an array of values. 

How do i correctly format my array of 2d tensors so they can be used as an input in the following example?

```
                    NDArray inputs = ValidationInputData( "ToFeatures" );
                    NDArray Outputs = ValidationOutputData();
                    FeedItem F1 = new FeedItem( ToInput, inputs );
                    FeedItem F2 = new FeedItem( Output, Outputs );

                    var ValidationResult = sess.run( tf.reduce_sum(tf.pow(FinalOutput - Output, 2.0f)) / (2.0f * ValidationData.Count), F1,F2);

                    Console.WriteLine($"Validation Results {ValidationResult}");
```

Is there an example of preparing the data to test a graph which takes a 2d tensor as an input?

**Comment by Oceania2018** [comment]
Try `tf.convert_to_tensor`.

**Comment by Rleahy** [comment]
What object type should i be passing to the FeedItem?

I'm trying to create a Tensor[] array then populating each slot on the array with tf.convert_to_tensor then passing that to the feed item. 
```

    Tensor[] NDValidationData;

        NDValidationData = new Tensor[ValidationData.Count];

        for(int i=0;i<ValidationData.Count;i++){

            NDValidationData[i] = tf.convert_to_tensor(  ValidationData[i].Inputs[InputSet] );
        
        }

        return NDValidationData;
```

{System.NotImplementedException: _run subfeed   at Tensorflow.BaseSession._run(Object fetches, FeedItem[] feed_dict)   at DeepLearningNetwork.TrainNetwork() in /BeliefRefinery/classes/StoredData/DeepLearningNetwork.cs:line 134   at TrainNeuralNetwork.CalculateAttribute() in /BeliefRefinery/classes/Attributes/TrainNeuralNetwork.cs:line 21   at CalculatedAttribute.TryCalculateAttribute() in /BeliefRefinery/classes/framework/CalculatedAttribute.cs:line 149}

**Comment by henon** [comment]
If I understand you correctly, you want to feed a 2D array into a tensor, right? 

Have you checked out my NeuralNetXor example? see NeuralNetXor.cs

This is the 2d data:
```csharp
NDArray data= new float[,]
            {
                {1, 0 },
                {1, 1 },
                {0, 0 },
                {0, 1 }
            };
```

This is the tensor:
```csharp

            var features = tf.placeholder(tf.float32, new TensorShape(4, 2));
```

and this is how the data is fed into the run method:

```csharp
var result = sess.run(new ITensorOrOperation[] { train_op, gs, loss }, new FeedItem(features, data), new FeedItem(labels, y_));

```

Hope this helps.

**Comment by Oceania2018** [comment]
@Rleahy That's a good example though.

**Comment by Rleahy** [comment]
@henon Thanks for the example however its not quite what I'm struggling with

Using the Liner Regressions example as a reference. 

```
 for (int epoch = 0; epoch < training_epochs; epoch++)
                {
                    foreach (var (x, y) in zip<float>(train_X, train_Y))
                    {
//Here Tensors are added to the feedItem in a loop one one tensor at a time then executed.
                        sess.run(optimizer, 
                            new FeedItem(X, x),
                            new FeedItem(Y, y));
                    }

                    // Display logs per epoch step
                    if ((epoch + 1) % display_step == 0)
                    {
//Here multiple data points are passed at once the array seems to contain the tensor values of all data points used to calculate the accuracy of the model. 
                        var c = sess.run(cost, 
                            new FeedItem(X, train_X),
                            new FeedItem(Y, train_Y));
                        Console.WriteLine($"Epoch: {epoch + 1} cost={c} " + $"W={sess.run(W)} b={sess.run(b)}");
                    }
}
```

What i need is an example of how to build and format the train_Y and train_X arrays in the above example if there input is a 2D tensor, rather than a single value.

**Comment by henon** [comment]
for linear regression train_X is not 2D, it is 1D. Also, the data is passed as a NumSharp NDArray into the FeedItem.

This is from our Linear regression example:
```csharp
            NDArray train_X = np.array(3.3f, 4.4f, 5.5f, 6.71f, 6.93f, 4.168f, 9.779f, 6.182f, 7.59f, 2.167f,
             7.042f, 10.791f, 5.313f, 7.997f, 5.654f, 9.27f, 3.1f);
            NDArray train_Y = np.array(1.7f, 2.76f, 2.09f, 3.19f, 1.694f, 1.573f, 3.366f, 2.596f, 2.53f, 1.221f,
                         2.827f, 3.465f, 1.65f, 2.904f, 2.42f, 2.94f, 1.3f);
```

you could also write it like this (using implicit conversion)
```csharp
NDArray train_X = new float[] { 3.3f, 4.4f, 5.5f, 6.71f, 6.93f, 4.168f, 9.779f, 6.182f, 7.59f, 2.167f,
             7.042f, 10.791f, 5.313f, 7.997f, 5.654f, 9.27f, 3.1f };
```

**Comment by Oceania2018** [comment]
2D array can be `NDArray train_X = new float[,]{{1f, 2f}, {3f, 4f}}`.

**Comment by Rleahy** [comment]
@henon @Oceania2018  Thanks again for the advice I'm sorry I'm not been clear enough with my question.  

I'm trying to reproduce the linear regression example with more complex inputs.

If X or Y in the example took inputs of a more complex shape such as (3,2) or (3,2,2) rather than just of shape(1) how would that change the away NDArrays are defined for testing. 

So if the liner regression example at lines 38 and 39
```
var X = tf.placeholder(tf.float32);
var Y = tf.placeholder(tf.float32);
```
was changed to

```
var X = tf.placeholder(tf.float32,new TensorShape(3,2));
var Y = tf.placeholder(tf.float32,new TensorShape(3,2,2));
```

How would that change the train_X and train_Y NDArrays that are used to calculate the training cost on line 88 of the example?

```
      var training_cost = sess.run(cost,
                    new FeedItem(X, train_X),
new FeedItem(Y, train_Y));
```

I've tried defining an NDArray of shape(10,3,2) to represent 10 sample inputs to X but it throws an out of bounds exception I'm hoping for an example of the correct way to do it.

**Comment by Oceania2018** [comment]
Have you checked the `KMeansClustering` example? 
If it doesn't help you out. You might have to provide the python code, then we can debug it.

**Comment by henon** [comment]
ok, I understand now what you mean. check out this link [tf.placeholder](https://www.tensorflow.org/api_docs/python/tf/placeholder)

if you do not specify the shape you can feed data of any shape. if you specify it you must feed data in exactly that shape. that is why feeding an array of shap (10,3,2) doesn't work when you define the placeholder with shape (3, 2, 2)

I don't know if it is possible to feed in multiple items with complex shapes but the analogy to feeding a 1D array instead of a scalar when applied to your example would mean you have to add a dimension to feed multiples of that complex shape. i.e (with a placeholder that has no shape defined) feed an array of (10, 3, 2, 2) to feed 10 items of shape (3, 2, 2). that would be my understanding.

**Comment by henon** [comment]
PS: the extra dimension is added at the front because in tensorflow dimension order is major-to-minor (akin to row-major at rank 2). See [Shapes and Layout](https://www.tensorflow.org/xla/shapes)

**Comment by Rleahy** [comment]
@henon  Yes that's what i'm asking :)

However i've already tried the solution your suggesting and it results in a runtime error. 

I have rewritten and cut down the liner regression example to demonstrate that error

```
  public class LinearRegression
    {
    

        // Parameters
        float learning_rate = 0.01f;
        NDArray train_X, train_Y;
        int n_samples;

        public void Run()
        {

            PrepareData();

            var X = tf.placeholder(tf.float32, new TensorShape(3,2));
            var Y = tf.placeholder(tf.float32);
            var W = tf.Variable(-0.06f, name: "weight");
            var b = tf.Variable(-0.73f, name: "bias");
            var pred = tf.add(tf.multiply(X, W), b);
            var cost = tf.reduce_sum(tf.pow(pred - Y, 2.0f)) / (2.0f * n_samples);
            var optimizer = tf.train.GradientDescentOptimizer(learning_rate).minimize(cost);
            var init = tf.global_variables_initializer();
            Session sess = tf.Session();
            sess.run(init);

            // Fit all training data runs without issue
            for(int i=0;i<n_samples;i++)
            {

                sess.run(optimizer, 
                    new FeedItem(X, train_X[i]),
                    new FeedItem(Y, train_Y[i]));
                    Console.WriteLine($"TrainingIteration: {i + 1}");
            }

            
            //error occures here
            //An unhandled exception of type 'System.Exception' occurred in TensorFlow.NET.dll: 'Incompatible shapes: [4,3,2] vs. [4][[{{node sub}}]]'
            var c = sess.run(cost, 
            new FeedItem(X, train_X),
            new FeedItem(Y, train_Y));

            Console.WriteLine("Finished executing");
 

        }

        private void PrepareData()
        {
            //setup array for 4 datapoints of shape (3,2) values unimportant for now as we simply want get the code to execute. 
            train_X = new NDArray(typeof(float));
            train_X = train_X.arange(4*3*2).reshape(4,3,2);
            //4 outputs to match inputs
            train_Y = np.array(1.7f, 2.76f, 2.09f, 3.19f);

            n_samples = train_X.shape[0];
        }
    
    }
```

Add the class to your project then create an instance and call run() in program main. 

Can you tell me if this is a bug or have i made a mistake setting up my arrays? If i have made a mistake how should i correct it?

**Comment by henon** [comment]
This one is easy to spot, no need to execute it. As I said before 

`var X = tf.placeholder(tf.float32, new TensorShape(3,2));`

must not be defined with a shape if you plan to feed it this way, so try this `var X = tf.placeholder(tf.float32);`. 

By the way, you can easily execute the same in Python and see what errors tensorflow throws in Python. Only if they are different, it is a bug in TensorFlow.NET.

**Comment by Rleahy** [comment]
How should i evaluate the cost of the model if one of my placeholders uses a TensorShape?

The model i'm producing has a large number of high rank tensors as inputs, defining each input as an individual value will quickly become a huge mess. 

If i wanted to stick with the shape (3,2) input how would i rewrite the

**Comment by henon** [comment]
you can do that (looping in your program). you can also use loops within tensorflow. but this question is best put to the tensorflow community. My knowledge is not detailed enough to answer that.
